File: Algorithms/234.回文链表/isPalindrome.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# 方法1：栈，时间复杂度O(n)，空间复杂度O(n)
# 692 ms
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        length = 0
        p = head
        while p:
            length += 1
            p = p.next

        stack = []
        p = head
        for i in range(length // 2):
            stack.append(p.val)
            p = p.next
        if length % 2 == 1:
            p = p.next
        while p:
            if stack[-1] != p.val:
                return False
            else:
                stack.pop()
                p = p.next
        return True

File: 程序员面试金典/2.6/isPalindrome.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null)
            return true;
        int length = lengthOfList(head);
        Stack<Integer> stack = new Stack<>();
        ListNode p = head;
        for (int i = 0; i < length / 2; ++i) {
            stack.push(p.val);
            p = p.next;
        }
        if (length % 2 == 1)
            p = p.next;
        for (int i = 0; i < length / 2; ++i) {
            if (p.val != stack.pop())
                return false;
            p = p.next;
        }
        return true;
    }

    private int lengthOfList(ListNode head) {
        int length = 0;
        while (head != null) {
            ++length;
            head = head.next;
        }
        return length;
    }

}
